### PR TITLE
Use program path as fallback location for eval file (option2)

### PR DESCRIPTION
**Pull request by snicolet**
This patch uses the program path (ie the the directory where the Stockfish
executable resides) as a fallback location to find the NNUE EvalFile. This
makes it possible to use NNUE eval files located in the Stockfish directory,
even in GUI which change the working directory under our feet.

No functional change.

### CODE REVIEW

**Comment by vondele** [comment]
I'm not particularly a fan of this, if at all possible we should avoid adding command line options. This doesn't make it easier for the users.

**Comment by snicolet** [comment]
Hmmm? This does not add command line options, does it?

**Comment by mstembera** [comment]
Maybe I'm missing something but I don't think this is adding any command line options.

**Comment by vondele** [comment]
sorry to quick a read, confused by `CommandLine::...`

**Comment by vondele** [comment]
I'm wondering how this works with paths that include non-ascii characters? I assume this is all escaped and works?

**Comment by snicolet** [comment]
Tested and working on Mac OS X (which have Terminal and command line utils which understand UTF-8):

workingDirectory =  /Users/stephane/日の出
argv0            =  ./stockfish
binaryDirectory  =  /Users/stephane/日の出/

**Comment by vondele** [comment]
sunrise :-)

**Comment by vondele** [comment]
works on linux too.

**Comment by snicolet** [comment]
sun

**Comment by vondele** [comment]
concerning the length of the path, 4k is pretty reasonable, seems like 32k is a thing on windows, and some variants on linux will be unlimited:
https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#maximum-path-length-limitation
https://www.gnu.org/software/libc/manual/html_mono/libc.html#Limits-for-Files

**Comment by snicolet** [comment]
Yes, I have seen about the infinite paths lengths, that's why I have checked the return pointer of getcwd. It is difficult to get the program directory in 100% of the corner cases (there are entire libraries written for that), but my opinion is that since we use this as a fallback option, it is sufficient to write code covering 99.9% of our uses.

**Comment by vdbergh** [comment]
If NULL is returned by getcwd and `errno==ERANGE` one could just retry with double the buffer size.

This is what the manual recommends.

**Comment by snicolet** [comment]
@Rocky640 @noobpwnftw

If you still have access to Visual C++, would it be possible to test this for a Windows build? 
Both running from the MS-DOS command, and if possible in a GUI similar to Fritz
(ie without magical support for nnue files).

Thanks :-)

File: src/misc.cpp [hunk 1]
```diff
@@ -589,3 +590,64 @@ void bindThisThread(size_t idx) {
...
+    if (pos == std::string::npos)
+        binaryDirectory = "." + pathSeparator;
+    else
+        binaryDirectory.resize(pos + 1);
+
+    // pattern replacement: "./" at the start of path is replaced by the working directory
+    if (binaryDirectory.find("." + pathSeparator) == 0)
+        binaryDirectory.replace(0, 1, workingDirectory);
+
+    std::cerr << "workingDirectory =  " << workingDirectory << std::endl;
```
Comment (vondele) [inline src/misc.cpp:647]: I assume that these std::cerr are only intended for debugging, right?

File: src/evaluate.cpp [hunk 2]
```diff
@@ -23,26 +23,33 @@
...
 
 namespace Eval {
 
   bool useNNUE;
-  std::string eval_file_loaded="None";
+  std::string eval_file_loaded = "None";
 
   void init_NNUE() {
 
+    std::vector<std::string> dirs = { "" , CommandLine::binaryDirectory };
```
Comment (vondele) [inline src/evaluate.cpp:43]: what about adding something as in https://github.com/official-stockfish/Stockfish/issues/3030#issuecomment-679107572 ?

File: src/misc.cpp (hunk 1 above)
Comment (snicolet) [reply src/misc.cpp:647]: Yep

File: src/evaluate.cpp (hunk 2 above)
Comment (snicolet) [reply src/evaluate.cpp:43]: I am working on it

Comment (snicolet) [reply src/evaluate.cpp:43]: See https://github.com/official-stockfish/Stockfish/pull/3070

**Comment by vondele** [comment]
I'll close that in favor of https://github.com/official-stockfish/Stockfish/pull/3070. Thanks for the PR!

**Pull request closed without merging**